### PR TITLE
[JENKINS-62496] Convert arrow pngs to CSS

### LINKED ISSUE #24: [JENKINS-62496] - Arrow icons don't work well
Arrows are invisible as they are black in most cases:

image

See https://issues.jenkins-ci.org/browse/JENKINS-62496

It would be good to change them to use CSS in core instead

### PR DESCRIPTION
**Pull request by EstherAF**
See [JENKINS-62496](https://issues.jenkins-ci.org/browse/JENKINS-62496).

Fixes https://github.com/jenkinsci/dark-theme/issues/24

Convert arrows used to display contextual menus, from background images to CSS triangles.

Screenshots of the change:
<details>

<summary>Down arrow</summary>

**Before:**
* Breadcrumbs: 
Jenkins-down-before
* Build Execution:
build-execution-down-before
* Build Queue:
build-queue-down-before
* Jobs:
job-down-before

**After:**
* Breadcrumbs: 
jenkins-down-after
* Build Execution:
build-execution-down-after
* Build Queue:
build-queue-down-after
* Jobs:
job-down-after

</details>

<details>
<summary>Right arrow</summary>

**Before** 
* Breadcrumbs
breadcrumbs-right-before

* Submenu
Screenshot from 2020-06-07 23-26-05

**After** 
* Breadcrumbs
breadcrumbs-right-after

* Submenu
Screenshot from 2020-06-07 23-30-55

</details>

### Proposed changelog entries

* Entry 1: JENKINS-62496 Arrow icons changed from images to CSS

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@timja @fqueiruga

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/resources/lib/layout/breadcrumbs.css
Comment (oleg-nenashev) [inline core/src/main/resources/lib/layout/breadcrumbs.css:109]: Will need update after https://github.com/jenkinsci/jenkins/pull/4752 is merged so that we do not hardcode colors

File: core/src/main/resources/lib/layout/breadcrumbs.css
Comment (oleg-nenashev) [inline core/src/main/resources/lib/layout/breadcrumbs.css:131]: Needs a variable after https://github.com/jenkinsci/jenkins/pull/4752

File: core/src/main/resources/lib/layout/breadcrumbs.css
Comment (oleg-nenashev) [inline core/src/main/resources/lib/layout/breadcrumbs.css:137]: Needs a variable after https://github.com/jenkinsci/jenkins/pull/4752

**Review by oleg-nenashev — COMMENTED** [review]
Looks good overall, thanks! I think we need to wait until https://github.com/jenkinsci/jenkins/pull/4752 is merged and then adds CSS variables for it

File: core/src/main/resources/lib/layout/breadcrumbs.css
Comment (fqueiruga) [reply core/src/main/resources/lib/layout/breadcrumbs.css:109]: And I think the variable should be `--text-color`.

File: core/src/main/resources/lib/layout/breadcrumbs.css
Comment (fqueiruga) [reply core/src/main/resources/lib/layout/breadcrumbs.css:131]: Ditto

File: core/src/main/resources/lib/layout/breadcrumbs.css
Comment (fqueiruga) [reply core/src/main/resources/lib/layout/breadcrumbs.css:137]: Same as above

**Review by timja — COMMENTED** [review]
I've played around with this with variables and I got it working, I'll send a PR here once the variables PR is merged,

All icons that I found except for one are working (the Manage Jenkins sub menu):

image


This appears to be the CSS https://github.com/jenkinsci/jenkins/blob/0e2fd168103dbdae015a43ea48e70e8816912eac/war/src/main/webapp/scripts/yui/menu/assets/menu.css#L409

Although I would suggest overriding it in the yui-compatibility less file: https://github.com/jenkinsci/jenkins/blob/49e501f37db848e69de31607a43df79bab4e77b0/war/src/main/less/base/yui-compatibility.less

**Comment by timja** [comment]
@EstherAF PR for the dark mode changes on the breadcrumbs: https://github.com/EstherAF/jenkins/pull/9

**Comment by EstherAF** [comment]
> All icons that I found except for one are working (the Manage Jenkins sub menu):

I'm looking at it. Do you know where the class `yuimenubaritem-hassubmenu` is being used? I've only found usages of the `yuimenuitem-hassubmenu`.

**Comment by timja** [comment]
> hassubmenu

can't see it, took me quite awhile to even find the css and selector that was causing the above issue

**Comment by varyvol** [comment]
@EstherAF it seems like the new arrows have a lighter dark than they previously had? I'm not against it, I just want to check it's something we've noticed and are fine with it.

Other than that, it looks good!

**Comment by EstherAF** [comment]
> @EstherAF it seems like the new arrows have a lighter dark than they previously had? I'm not against it, I just want to check it's something we've noticed and are fine with it.

Yes! It is because now they are using the same color than the text, with the css variable. Before they were using just black (IIRC). 

What do you think? I prefer this so the color palette used is simpler, and it is more integrated with the text of the link.

**Comment by varyvol** [comment]
@EstherAF it makes sense to me and I agree. As I said, I just wanted to note it just in case.

**Review by fqueiruga — APPROVED** [review]

**Comment by EstherAF** [comment]
> All icons that I found except for one are working (the Manage Jenkins sub menu)

@timja I've just pushed a fix for that. 

I've faced a little problem here that I don't know how to solve it. The arrow with CSS is generated as another element inside the menu item, as a sibling of the menu link (the text). This is causing that the menu link is not 100% wide anymore, and when focused the borders are displayed and you can see that the size is different.
Here it is more clear, look at the red border:

Before
Screenshot from 2020-06-07 23-26-05
After. There is a space at the right of the red border.
Screenshot from 2020-06-07 23-30-55

It only happens when the menu item has the arrow at the right (when it has a submenu). In the other items, the border on focus remains like before.

I don't know if this is a big deal, WDYT?

**Review by timja — APPROVED** [review]
I've tried it out and wasn't able to come up with anything better,

thanks!

File: war/src/main/less/base/yui-compatibility.less
Comment (fqueiruga) [inline war/src/main/less/base/yui-compatibility.less:165]:
```suggestion
```

File: war/src/main/less/base/yui-compatibility.less
Comment (fqueiruga) [inline war/src/main/less/base/yui-compatibility.less:179]:
```suggestion
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    right: 0.75em;
    display: inline-block;
```

File: war/src/main/less/base/yui-compatibility.less
Comment (fqueiruga) [inline war/src/main/less/base/yui-compatibility.less:181]:
```suggestion
}

```

File: war/src/main/less/base/yui-compatibility.less
Comment (fqueiruga) [inline war/src/main/less/base/yui-compatibility.less:175]:
```suggestion
    border-left: 0.35em solid #333;
    border-left: 0.35em solid var(--text-color);
    border-top: 0.35em solid transparent;
    border-bottom: 0.35em solid transparent;
```

**Review by fqueiruga — CHANGES_REQUESTED** [review]
I've suggested some changes that should fix the issue with the dropdown menu. Please test them locally befor committing :D

**Review by timja — APPROVED** [review]
Tested, works great!

**Review by fqueiruga — APPROVED** [review]

**Comment by timja** [comment]
Thanks for the PR!

This PR is now ready-for-merge, we will merge it after ~24 hours if there's no negative feedback

**Review by oleg-nenashev — APPROVED** [review]
Looks good, all comments are addressed. Thanks @EstherAF !

**Pull request merged**
